src/routing_table/mod.rs

```diff
@@ -662,16 +654,15 @@ impl<T: Binary + Clone + Copy + Debug + Default + Hash + Xorable> RoutingTable<T
             return OwnMergeState::AlreadyMerged;
         }
         for prefix in merge_details.sections.keys() {
-            let compatible_with_ours = self.our_prefix.is_compatible(prefix);
-            // This may be a section which has been merged from multiple sections currently still
-            // in our RT, so fix up our RT first.
-            if merge_details.merge_prefix.is_compatible(prefix) &&
-               !self.sections.contains_key(prefix) && !compatible_with_ours {
-                self.merge(prefix);
-            }
-            // Add an empty section in the table.
-            if !compatible_with_ours {
-                let _ = self.sections.entry(*prefix).or_insert_with(HashSet::new);
+            if *prefix == self.our_prefix || self.sections.contains_key(prefix) {
+                continue; // Already in our routing table.
+            } else if self.our_prefix.is_compatible(prefix) ||
```

Comment:
Should we also check that the prefix is a neighbour, and that after we're done there aren't any missing prefixes? Okay, `finish_merging...` adds missing prefixes, but if there were any that itself should be an error.